crates/bevy_core_pipeline/src/smaa/smaa.wgsl

```diff
@@ -0,0 +1,1122 @@
+const SMAA_SEARCHTEX_PACKED_SIZE: vec2<f32> = vec2(64.0, 16.0);
+
+#ifndef SMAA_DISABLE_CORNER_DETECTION
+const SMAA_CORNER_ROUNDING_NORM: f32 = f32(SMAA_CORNER_ROUNDING) / 100.0;
+#endif  // SMAA_DISABLE_CORNER_DETECTION
+
+//-----------------------------------------------------------------------------
+// Porting Functions
+
+#ifdef SMAA_BLENDING_WEIGHT_CALCULATION
+fn sample_edges_level_zero_offset(coord: vec2<f32>, offset: vec2<i32>) -> vec4<f32> {
+    return textureSampleLevel(
+        edges_texture,
+        edges_sampler,
+        coord + vec2<f32>(offset) * smaa_info.rt_metrics.xy,
+        0.0
+    );
+}
+#endif  // SMAA_BLENDING_WEIGHT_CALCULATION
+
+//-----------------------------------------------------------------------------
+// WGSL-Specific Functions
+
+// This vertex shader produces the following, when drawn using indices 0..3:
+//
+//  1 |  0-----x.....2
+//  0 |  |  s  |  . ´
+// -1 |  x_____x´
+// -2 |  :  .´
+// -3 |  1´
+//    +---------------
+//      -1  0  1  2  3
+//
+// The axes are clip-space x and y. The region marked s is the visible region.
+// The digits in the corners of the right-angled triangle are the vertex
+// indices.
+//
+// The top-left has UV 0,0, the bottom-left has 0,2, and the top-right has 2,0.
+// This means that the UV gets interpolated to 1,1 at the bottom-right corner
+// of the clip-space rectangle that is at 1,-1 in clip space.
+fn calculate_vertex_varyings(vertex_index: u32) -> VertexVaryings {
+    // See the explanation above for how this works
+    let uv = vec2<f32>(f32(vertex_index >> 1u), f32(vertex_index & 1u)) * 2.0;
+    let clip_position = vec2<f32>(uv * vec2<f32>(2.0, -2.0) + vec2<f32>(-1.0, 1.0));
+
+    return VertexVaryings(clip_position, uv);
+}
+
+fn texture_offset(coords: vec2<f32>, offset: vec2<i32>) -> vec2<f32> {
+    return coords + vec2<f32>(offset) * smaa_info.rt_metrics.xy;
+}
+
+//-----------------------------------------------------------------------------
+// Vertex Shaders
+
+#ifdef SMAA_EDGE_DETECTION
+
+/**
+ * Edge Detection Vertex Shader
+ */
+@vertex
+fn edge_detection_vertex_main(@builtin(vertex_index) vertex_index: u32) -> EdgeDetectionVaryings {
+    let varyings = calculate_vertex_varyings(vertex_index);
+
+    var edge_detection_varyings = EdgeDetectionVaryings();
+    edge_detection_varyings.position = vec4(varyings.clip_coord, 0.0, 1.0);
+    edge_detection_varyings.tex_coord = varyings.tex_coord;
+
+    edge_detection_varyings.offset_0 = smaa_info.rt_metrics.xyxy * vec4(-1.0, 0.0, 0.0, -1.0) +
+        varyings.tex_coord.xyxy;
+    edge_detection_varyings.offset_1 = smaa_info.rt_metrics.xyxy * vec4(1.0, 0.0, 0.0, 1.0) +
+        varyings.tex_coord.xyxy;
+    edge_detection_varyings.offset_2 = smaa_info.rt_metrics.xyxy * vec4(-2.0, 0.0, 0.0, -2.0) +
+        varyings.tex_coord.xyxy;
+
+    return edge_detection_varyings;
+}
+
+#endif  // SMAA_EDGE_DETECTION
+
+#ifdef SMAA_BLENDING_WEIGHT_CALCULATION
```

Comment:
They're necessary because otherwise there will be undefined references to uniforms that don't exist, since the per-stage uniforms are behind `#ifdef`s.